Bevy 0.16 update

RobertBrewitz opened the pull request:
Update implementation, examples and dependencies to work with bevy 0.16.

All examples seem to run fine.

RobertBrewitz commented:
The commit https://github.com/linebender/bevy_vello/pull/152/commits/4b2da2859157aebc2fcdcc62fe6e26104babce5b removes a lot of visibility code. Bevy 0.16 seem to have changed how visibility works and needs to be thought about thoroughly.

The migration documentation does not mention this, so I assume we have been using a low level API for this. The only mention of visibility in migration docs are here: https://bevyengine.org/learn/migration-guides/0-15-to-0-16/#remove-the-type-parameter-from-check-visibility-and-visibleentities

Ideas are welcomed!

StarArawn commented:
> The commit [fbf01ec](https://github.com/linebender/bevy_vello/commit/fbf01ec30a5b2c612ea1390ee563ea7b3547a08f) removes a lot of visibility code. Bevy 0.16 seem to have changed how visibility works and needs to be thought about thoroughly.
> 
> The migration documentation does not mention this, so I assume we have been using a low level API for this. The only mention of visibility in migration docs are here: https://bevyengine.org/learn/migration-guides/0-15-to-0-16/#remove-the-type-parameter-from-check-visibility-and-visibleentities
> 
> Ideas are welcome.

In my shoddy port I just removed the visibility stuff completely. The migration guide or docs mentioned that the system that runs to handle visibility is not type specific anymore so it should just work? I haven't seen any issues in my code, but I'm also not using the visibility component for a vello bevy scene.

RobertBrewitz commented:
> > The commit [fbf01ec](https://github.com/linebender/bevy_vello/commit/fbf01ec30a5b2c612ea1390ee563ea7b3547a08f) removes a lot of visibility code. Bevy 0.16 seem to have changed how visibility works and needs to be thought about thoroughly.
> > The migration documentation does not mention this, so I assume we have been using a low level API for this. The only mention of visibility in migration docs are here: https://bevyengine.org/learn/migration-guides/0-15-to-0-16/#remove-the-type-parameter-from-check-visibility-and-visibleentities
> > Ideas are welcome.
> 
> In my shoddy port I just removed the visibility stuff completely. The migration guide or docs mentioned that the system that runs to handle visibility is not type specific anymore so it should just work? I haven't seen any issues in my code, but I'm also not using the visibility component for a vello bevy scene.

Yeah, I removed it completely as well here :smiley: 

Anyway, if we want visibility and inherited visibility to work as intended, our extract functions need to be rewritten. If you try to revert commit `fbf01ec30a5b2c612ea1390ee563ea7b3547a08f` nothing renders because `visibility.get()` and `inherited_visibility.get()` is always return `false`.

simbleau commented on CHANGELOG.md:
```diff
@@ -15,6 +15,22 @@ You can find its changes [documented below](#080---2025-04-29).
 
+### Added
+
+- adds `tracing` crate as bevy removed built-in tracing macros
+
+### Changed
+
+- updates `bevy` to 0.16
+- updates `vello` to 0.5.0
+- updates `velato` to 0.6.0
+- updates `vello_svg` to 0.7.0
+- updates `parley` to 0.4.0
+
+### Removed
+
+- visibility logic removed pending new implementation
```
```suggestion
- Adds `tracing` crate as bevy removed built-in tracing macros

### Changed

- Updates `bevy` to 0.16
- Updates `vello` to 0.5.0
- Updates `velato` to 0.6.0
- Updates `vello_svg` to 0.7.0
- Updates `parley` to 0.4.0

### Removed

- Visibility logic removed pending new implementation
```

simbleau commented on src/integrations/text/font.rs:
```diff
@@ -54,8 +55,10 @@ impl VelloFont {
             let font_context = font_context.as_mut().unwrap();
 
             LOCAL_LAYOUT_CONTEXT.with_borrow_mut(|layout_context| {
+                // TODO: fourth argument is new in parley 0.4.0, should perhaps be defaulted to
```
What is the new argument?

RobertBrewitz replied:
The new argument is `quantize`.

>    /// Set `quantize` as `true` to have the layout coordinates aligned to pixel boundaries.
>    /// That is the easiest way to avoid blurry text and to receive ready-to-paint layout metrics.
>    ///
>    /// For advanced rendering use cases you can set `quantize` as `false` and receive
>    /// fractional coordinates. This ensures the most accurate results if you want to perform
>    /// some post-processing on the coordinates before painting. To avoid blurry text you will
>    /// still need to quantize the coordinates just before painting.

https://github.com/linebender/parley/blob/v0.4.0/CHANGELOG.md?plain=1#L31

> For an easy upgrade we recommend enabling it by setting `quantize` to `true` ..

Seems true is fine.

simbleau reviewed (commented):
I'm a little concerned about Visibility needing to be removed. Primarily how we are *supposed* to migrate. But overall looks good on first pass. While velato gets updated, I'd like to investigate more.

RobertBrewitz commented:
Beware that https://github.com/linebender/bevy_vello/pull/152/commits/f56f5362f9e102daa00ae8008a22395244f86315 is causing rendering to **not** work.

@simbleau

RobertBrewitz commented:
> I'm a little concerned about Visibility needing to be removed. Primarily how we are _supposed_ to migrate. But overall looks good on first pass. While velato gets updated, I'd like to investigate more.

Indeed, I am doing something else tonight but will take some time this week to look into how we can hook into Beyv's visibility again, must be something minor that is missing.

RobertBrewitz commented:
I will dig into this for a few hours today and see if I can get better informed about how the ViewVisibility actually works under the hood.

StarArawn commented on src/lib.rs:
```diff
@@ -51,12 +51,14 @@ pub struct VelloSceneBundle {
 /// A simple newtype component wrapper for [`vello::Scene`] for rendering.
 #[derive(Component, Default, Clone, Deref, DerefMut)]
-#[require(Transform, ViewVisibility)]
+#[require(Transform, Visibility, VisibilityClass)]
```
Bevy sprite also has this:
https://github.com/bevyengine/bevy/blob/e9418b3845c1ffc9624a3a4003bde66a2ad6566a/crates/bevy_sprite/src/sprite.rs#L19

RobertBrewitz commented:
I am still struggling understanding exactly how InheritedVisibility works at this time; Using that component in the extract query and calling `.get()` results in a false result.

RobertBrewitz commented:
Ok, view culling seems to be working, need to clean up the new view_culling example and perhaps rewrite it a bit to make more sense.

It's just a quick and dirty initial testing example.

RobertBrewitz commented:
woop woop, inherited visibility works again

RobertBrewitz commented:
The calculation to get the bounding box of lotties and svgs are either wrong or I'm using them wrong.

RobertBrewitz commented:
Still got work to do in the **view_culling example**, scale is not applied to the Aabb component right now.

simbleau commented:
I don't see the view_culling example

RobertBrewitz commented:
> I don't see the view_culling example

It should be under `examples/view_culling/src/main.rs` :monocle_face: :smiley:

RobertBrewitz commented on examples/view_culling/src/main.rs:
```diff
@@ -0,0 +1,219 @@
+        .insert(Aabb::default());
+}
+
+fn update_lottie_aab(
```
I would probably not want to run this on every update after cleaning up this code :)
